From 16a5f4949cb829cffd048477f0a0116928c38a10 Mon Sep 17 00:00:00 2001
From: "Cao,Bing Bu" <bingbu.cao@intel.com>
Date: Fri, 24 Apr 2015 15:37:09 +0800
Subject: [PATCH 385/462] atomisp2: NOT get bq info from a NULL stream

In camera3hal, the 3AThread may try to get dvs2_bq_resolution
info from stream which maybe is not created yet.
The driver should return error instead of crash.

Change-Id: I66352b36e464d2387b9816e30c3f66b7441b6f7d
Traced-On: #JIMINAN-28392
Signed-off-by: Cao,Bing Bu <bingbu.cao@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/52972
Approver: Jing Tao <jing.tao@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../media/pci/atomisp2/atomisp_driver/atomisp_cmd.c  | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index 4063c86..23a18d3 100755
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -2402,14 +2402,22 @@ int atomisp_set_array_res(struct atomisp_sub_device *asd,
 int atomisp_get_dvs2_bq_resolutions(struct atomisp_sub_device *asd,
 			 struct atomisp_dvs2_bq_resolutions *bq_res)
 {
-	struct ia_css_pipe_config *pipe_cfg =
-		&asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL]
+	struct ia_css_pipe_config *pipe_cfg = NULL;
+	struct ia_css_stream_config *stream_cfg = NULL;
+	struct ia_css_stream_input_config *input_config = NULL;
+
+	struct ia_css_stream *stream =
+		asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL].stream;
+	if (!stream) {
+		dev_warn(asd->isp->dev, "stream is not created");
+		return -EAGAIN;
+	}
+
+	pipe_cfg = &asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL]
 		.pipe_configs[CSS_PIPE_ID_VIDEO];
-	struct ia_css_stream_config *stream_cfg =
-		&asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL]
+	stream_cfg = &asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL]
 		.stream_config;
-	struct ia_css_stream_input_config *input_config =
-		&stream_cfg->input_config;
+	input_config = &stream_cfg->input_config;
 
 	if (!bq_res)
 		return -EINVAL;
-- 
1.9.1

